dot/rpc/modules/author.go

```diff
@@ -94,26 +116,83 @@ func NewAuthorModule(logger log.Logger, coreAPI CoreAPI, txStateAPI TransactionS
 	}
 }
 
-// InsertKey inserts a key into the keystore
-func (am *AuthorModule) InsertKey(r *http.Request, req *KeyInsertRequest, res *KeyInsertResponse) error {
-	keyReq := *req
+// HasSessionKeys checks if the keystore has private keys for the given session public keys.
+func (am *AuthorModule) HasSessionKeys(r *http.Request, req *HasSessionKeyRequest, res *HasSessionKeyResponse) error {
+	pubKeysBytes, err := common.HexToBytes(req.Data)
+	if err != nil {
+		return err
+	}
+
+	pkeys, err := scale.Marshal(pubKeysBytes)
+	if err != nil {
+		am.logger.Debug("err while scale encoding data", "err", err)
+		return err
+	}
+
+	data, err := am.coreAPI.DecodeSessionKeys(pkeys)
+	if err != nil {
+		am.logger.Debug("err while calling decode session keys runtime function", "err", err)
+		*res = false
+		return err
+	}
 
-	pkDec, err := common.HexToBytes(keyReq[1])
+	var decodedKeys *[]DecodedKey
+	err = scale.Unmarshal(data, &decodedKeys)
 	if err != nil {
 		return err
 	}
 
-	privateKey, err := keystore.DecodePrivateKey(pkDec, keystore.DetermineKeyType(keyReq[0]))
+	qtyCheck := 0
+	for _, key := range *decodedKeys {
+		if common.CheckAllBytesZero(key.Data) {
+			continue
+		}
+
+		encType := keystore.Name(key.Type[:])
+		ok, err := am.coreAPI.HasKey(common.BytesToHex(key.Data), string(encType))
+
+		if err != nil {
+			*res = false
+			return err
+		}
+
+		if !ok {
+			*res = false
+			return nil
+		}
+
+		qtyCheck++
+	}
+
+	if qtyCheck > 0 {
+		*res = true
+	} else {
+		*res = false
+	}
+
```

Comment:
the spec says `Returns true if all private keys could be found, false if otherwise.` so this should be true if `qtyCheck` equals the number of input public keys, false otherwise

Reply:
Logic updated, I removed the `qtyCheck` and if there is some problem in the iteration then quick return `false` otherwise, if all the iterations are completed then return `true`

Reply:
While testing against the khala-network script, when the gossamer runtime parse the request data into keys we got these keys:
![image](https://user-images.githubusercontent.com/17255488/127072248-cd572c03-ac22-4e98-a66b-8df95c276238.png)

The first key is correct and the node contains it, but the others doesn't make sense, then I ask in element chat about this behaviour -> https://matrix.to/#/!HzySYSaIhtyWrwiwEV:matrix.org/$1627341312134571bzhXc:matrix.org?via=matrix.parity.io&via=corepaper.org&via=matrix.org